cedar-policy-core/src/authorizer/partial_response.rs

```diff
@@ -0,0 +1,481 @@
+            Either::Left(self.satisfied_permit_ids())
+        } else {
+            // Otherwise it's the true forbids
+            Either::Right(self.satisfied_forbid_ids())
+        }
+    }
+
+    /// Returns the set of non-trivial (meaning more than just `true` or `false`) residuals expressions
+    pub fn nontrivial_residuals(&self) -> impl Iterator<Item = &Expr> {
+        self.residual_forbids
+            .values()
+            .chain(self.residual_permits.values())
+            .map(|(r, _)| r)
+    }
+
+    /// Returns every policy as a residual expression
+    pub fn all_residuals(&self) -> impl Iterator<Item = &Expr> {
+        self.all_permit_residuals()
+            .chain(self.all_forbid_residuals())
+            .map(|(_, (residual, _))| residual)
+    }
+
+    /// Returns all residuals expressions that come from [`Effect::Permit`] policies
+    fn all_permit_residuals(&self) -> impl Iterator<Item = (&PolicyID, (&Expr, &Annotations))> {
+        let trues = self
+            .satisfied_permits
+            .iter()
+            .map(|(id, a)| (id, (&self.true_expr, a)));
+        let falses = self
+            .false_permits
+            .iter()
+            .map(|(id, (_, a))| (id, (&self.false_expr, a)));
+        let nontrivial = self
+            .residual_permits
+            .iter()
+            .map(|(id, (r, a))| (id, (r, a)));
+        trues.chain(falses).chain(nontrivial)
+    }
+
+    /// Returns all residuals expressions that come from [`Effect::Forbid`] policies
+    fn all_forbid_residuals(&self) -> impl Iterator<Item = (&PolicyID, (&Expr, &Annotations))> {
+        let trues = self
+            .satisfied_forbids
+            .iter()
+            .map(|(id, a)| (id, (&self.true_expr, a)));
+        let falses = self
+            .false_forbids
+            .iter()
+            .map(|(id, (_, a))| (id, (&self.false_expr, a)));
+        let nontrivial = self
+            .residual_forbids
+            .iter()
+            .map(|(id, (r, a))| (id, (r, a)));
+        trues.chain(falses).chain(nontrivial)
+    }
+
+    /// Return the residual for a given [`PolicyID`], if it exists in the response
+    pub fn get(&self, id: &PolicyID) -> Option<&Expr> {
+        self.get_true(id)
+            .or_else(|| self.get_false(id).or_else(|| self.get_residual(id)))
+    }
+
+    fn get_true(&self, id: &PolicyID) -> Option<&Expr> {
+        self.satisfied_permits
+            .get(id)
+            .or_else(|| self.satisfied_forbids.get(id))
+            .map(|_| &self.true_expr)
+    }
+
+    fn get_false(&self, id: &PolicyID) -> Option<&Expr> {
+        self.false_permits
+            .get(id)
+            .or_else(|| self.false_forbids.get(id))
+            .map(|_| &self.false_expr)
+    }
+
+    fn get_residual(&self, id: &PolicyID) -> Option<&Expr> {
+        self.residual_permits
+            .get(id)
+            .or_else(|| self.residual_forbids.get(id))
+            .map(|(r, _)| r)
+    }
```

Comment:
these methods might want doc comments.  I suppose they get the residual for the given id, but only if the residual is true/false/nontrivial respectively, otherwise `None` if the policy doesn't exist or has a different residual type?